src/Compilers/CSharp/Test/Semantic/Semantics/RefExtensionMethodsTests.cs

```diff
@@ -0,0 +1,438 @@
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref readonly this int p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        int x = 5;
+        x.PrintValue();
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { SystemCoreRef }, expectedOutput: "5");
+        }
+
+        [Fact]
+        public void RefExtensionMethodsReceiverTypes_ValueTypes()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref this int p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        int x = 5;
+        x.PrintValue();
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { SystemCoreRef }, expectedOutput: "5");
+        }
+
+        [Fact]
+        public void RefExtensionMethodsReceiverTypes_ReferenceTypes()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref this string p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        string x = ""test"";
+        x.PrintValue();
+    }
+}";
+
+            CreateCompilationWithMscorlibAndSystemCore(code).VerifyDiagnostics(
+                // (4,24): error CS8414: The receiver of the reference extension method 'PrintValue' must be a value type or a generic type constrained to one.
+                //     public static void PrintValue(ref this string p)
+                Diagnostic(ErrorCode.ERR_RefExtensionMustBeValueTypeOrConstrainedToOne, "PrintValue").WithArguments("PrintValue").WithLocation(4, 24));
+        }
+
+        [Fact]
+        public void RefExtensionMethodsReceiverTypes_UnconstrainedGenericTypes()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue<T>(ref this T p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        string x = ""test"";
+        x.PrintValue();
+    }
+}";
+
+            CreateCompilationWithMscorlibAndSystemCore(code).VerifyDiagnostics(
+                // (4,24): error CS8414: The receiver of the reference extension method 'PrintValue' must be a value type or a generic type constrained to one.
+                //     public static void PrintValue(ref this string p)
+                Diagnostic(ErrorCode.ERR_RefExtensionMustBeValueTypeOrConstrainedToOne, "PrintValue").WithArguments("PrintValue").WithLocation(4, 24));
+        }
+
+        [Fact]
+        public void RefExtensionMethodsReceiverTypes_StructConstrainedGenericTypes()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue<T>(ref this T p) where T : struct
+    {
+        System.Console.WriteLine(p);
```

Comment:
conditional access operators cannot be used with value types? I'll add a test to verify this is propagated properly.
Will add tests for recursion as well.

---
In reply to: [120229786](https://github.com/dotnet/roslyn/pull/19703#discussion_r120229786) [](ancestors = 120229786)